replay attack prevention (#54)
vcstuff/draft-ietf-oauth-attestation-based-client-auth: draft-ietf-oauth-attestation-based-client-auth.md

The guidance provided by {{RFC7519}} and {{RFC8725}} applies.

> ## Replay Attack Detection

Reviewer: ```suggestion
## Replay Attack Detection

The following mechanisms exist within this client authentication method in order to allow an authorization server to detect replay attacks for presented client attestation PoPs:
```

> Replay attack prevention MUST be considered by the protocols using this client authentication method. Possible mechanisms comprise:

Reviewer: Should we make this into a sub heading like `Replay Attack Detection`?

> Replay attack prevention MUST be considered by the protocols using this client authentication method. Possible mechanisms comprise:

Reviewer: ```suggestion
## Replay Attack Detection
```

- The client uses "jti" (JWT ID) claims for the Client Attestation PoP JWT and the authorization server maintains a list of used (seen) "jti" values for the time of which the JWT would be considered valid based on the applicable "exp" claim. If any Client Attestation PoP JWT would be replayed, the authorization server would recognize the "jti" and respond with an authentication error.
- The authorization server provides a nonce for the particular transaction and the client uses it for the "nonce" claim in the Client Attestation PoP JWT. The authorization server validates that the nonce matches for the transaction. This approach may require an additional roundtrip in the protocol. The authorization server MUST ensure that the nonce provides sufficient entropy.
- The authorization server may expect the usage of a nonce in the Client Attestation PoP JWT, but instead of providing the nonce explicitly, the client may implicitly reuse an existing artefact, e.g. the authorization code. The authorization server MUST ensure that the nonce provides sufficient entropy.

> The approach using a nonce explicitly provided by the authorization server is usually prefered, however support by the authorization server is OPTIONAL. The "jti" method is mandatory and hence acts as a default fallback.

Reviewer: ```suggestion
The approach using a nonce explicitly provided by the authorization server provides stronger replay attack detection guarantees, however support by the authorization server is OPTIONAL to simplify mandatory implementation requirements. The "jti" method is mandatory and hence acts as a default fallback.
```
